### PR TITLE
memory efficient hybrid search

### PR DESCRIPTION
**Pull request by junha-l**
Improve memory efficiency of GPU Hybrid Search. 

## Algorithm
It is based on a very simple algorithm. Moreover, it doesn't require two passes of distance calculation.
Please refer ```WriteNeighborsHybridKernel``` function in ```cpp/open3d/core/nns/FixedRadiusSearch.h```

#### Pseudocode 
For each query point,
- Initialize ```array``` with size ```{k}```.
- Initialize ```max_dist = INF```, ```max_index = INF```, and ```count = 0```. 
- Iterate over dataset points within nearby hash bin, and calculate distance, ```d```.
- If ```count < k```, then ```array[count] = d,  increase count, update max_dist, max_index```: O(1)
- else  if ```d < max_dist```, then ```array[max_index] = d, find new max_dist, max_index```: O(k)
- After finish iteration, sort the results: O(k^2)

The computational complexity increases as ```k``` increases. 
So when ```k << num_dataset```, it is more computationally efficient than the previous version as well.

## Benchmark Result
The memory usage is significantly reduced, approx 200MB. 

image

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3073)

### CODE REVIEW

**Comment by reyanshsolis** [comment]
Hi @junha-l, the memory issue is resolved, however, there were some performance concerns, so I have posted some results from my testing on PR #3064 [comment link](https://github.com/intel-isl/Open3D/pull/3064#issuecomment-784434312).
Thanks!

File: cpp/open3d/core/nns/FixedRadiusIndex.h [hunk 1]
```diff
@@ -105,12 +105,23 @@ class NeighborSearchAllocator {
         *ptr = indices.GetDataPtr<int64_t>();
     }
 
+    void AllocIndices(int64_t** ptr, size_t num, int64_t value) {
+        indices = Tensor::Full({int64_t(num)}, value, Dtype::Int64, device_);
```
Comment (yxlao) [inline cpp/open3d/core/nns/FixedRadiusIndex.h:109]: rename to `indices_`

File: cpp/open3d/core/nns/FixedRadiusIndex.h [hunk 2]
```diff
@@ -105,12 +105,23 @@ class NeighborSearchAllocator {
...
+        *ptr = indices.GetDataPtr<int64_t>();
+    }
+
     void AllocDistances(T** ptr, size_t num) {
         distances =
                 Tensor::Empty({int64_t(num)}, Dtype::FromType<T>(), device_);
         *ptr = distances.GetDataPtr<T>();
```
Comment (yxlao) [inline cpp/open3d/core/nns/FixedRadiusIndex.h:116]: rename to `distances_`

File: cpp/open3d/core/nns/FixedRadiusSearch.cu [hunk 3]
```diff
@@ -512,28 +512,209 @@ void WriteNeighborsIndicesAndDistances(
...
-                j >= num_neighbors ? -1 : prev_distances[indices_offset + j];
+    Vec3<T> query_pos(query_points[query_idx * 3 + 0],
+                      query_points[query_idx * 3 + 1],
+                      query_points[query_idx * 3 + 2]);
+    Vec3<int> voxel_index = ComputeVoxelIndex(query_pos, inv_voxel_size);
+    int hash = SpatialHash(voxel_index) % hash_table_size;
+
+    int bins_to_visit[8] = {hash, -1, -1, -1, -1, -1, -1, -1};
+
+    for (int dz = -1; dz <= 1; dz += 2)
```
Comment (yxlao) [inline cpp/open3d/core/nns/FixedRadiusSearch.cu:546]: let's add the full `{}` for the for loops

**Review by yxlao — APPROVED** [review]

File: cpp/open3d/core/nns/FixedRadiusIndex.h (hunk 1 above)
Comment (junha-l) [reply cpp/open3d/core/nns/FixedRadiusIndex.h]: Done.

File: cpp/open3d/core/nns/FixedRadiusIndex.h (hunk 2 above)
Comment (junha-l) [reply cpp/open3d/core/nns/FixedRadiusIndex.h]: Done.

File: cpp/open3d/core/nns/FixedRadiusSearch.cu (hunk 3 above)
Comment (junha-l) [reply cpp/open3d/core/nns/FixedRadiusSearch.cu]: Done.

**Review by benjaminum — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3073)* status: 0 of 5 files reviewed, 4 unresolved discussions (waiting on @junha-l and @yxlao)

___
*[cpp/open3d/core/nns/FixedRadiusSearch.h, line 256 at r2](https://reviewable.io/reviews/intel-isl/open3d/3073#-MUSBy6jBWm4vMGr286O:-MUSBy6jBWm4vMGr286P:bvuapxr) ([raw file](https://github.com/intel-isl/open3d/blob/98419588ae687c22cd634cd78a375284e1188ab7/cpp/open3d/core/nns/FixedRadiusSearch.h#L256)):*
> ```objc
> ///        array must be equal to the number of points.
> ///
> /// \param metric    One of L1, L2, Linf. Defines the distance metric for the
> ```

The function does not have this parameter.

**Pull request merged**